### ISSUE TITLE
[Bug]: Prefetch seems incompatible with a shared exporter

### ISSUE THREAD

**Issue by osd42** [solacecommunity/solace-prometheus-exporter#67]
### Bug Description

Hello, 
Using 1.5.1, I am testing the prefetch feature but I am not able to get the expected results for some of our setups. This is a great feature but it is seems limited to one use-case:

As soon as you activate the option "prefetchInterval = 30s", all scrapes will query the same "scrapeURI" (defined in the config file or as an environment variable)) even if you override the GET parameter "&scrapeURI="

If you activate the option without specifying the scrapeURI variable, you get an error
```text
solace_up{error="Post \"none/SEMP\": unsupported protocol scheme \"\""} 0
```

So I managed to test the feature using a "one node only" configuration (as we monitor solace cloud nodes with dedicated exporters per node). It is so powerfull, it fixes all my timeouts issues (it has been days working just on timeouts) !

Another Bug, if you run the feature as expected (single node monitoring): during the loading of the container we can see in the logs the prefetched requests for each Scope involved in the different custom endpoints. But if you query one of the endpoints during the first prefetch cache is not finished in background, you will get an empty response with HTTP code 200 and no trace of the call in the logfile. (It took me a lot of time to understand I just need to wait as we have sometimes multiple scopes that get 15 seconds to retrieve on an overloaded HA broker).
Perhaps you should send a "503" HTTP code (under maintenance) or the request should wait for the cache no matter how long it takes and always send a solace_up{} at least? 

best regards !

### Expected Behavior

I suppose the assomption to use the config file tuples of [node * scopes in custom_endpoints] to create a cache is too much limited.

If you use a Solace Prometheus Exporter to monitor several nodes, and different types of nodes from different clusters (where you need the most prefetch), we should use the tuples coming from requests. ex: [ /brokerha&scrapeURI=https://broker1.io:443, /brokerha2&scrapeURI=https://broker1.io:443, /brokerha&scrapeURI=https://broker2.io:443, /brokerha2&scrapeURI=https://broker2.io:443, /cloud&scrapeURI=https://mr-connection-545az454daz.messaging.solace.cloud:943, /cloud2&scrapeURI=https://mr-connection-545az454daz.messaging.solace.cloud:943 ] etc...

I use different endpoints for different type of targets (brokerha, clusters, cloud). So with the current solution I have to: 
- create 1 exporter per node
- create 1 config per exporter (1 per type of solace nodes)

With the tuples from the external requests, it should cover all use-cases (shared and dedicated, and only used custom endpoints, not all the endpoints for all the nodes).

### Steps to Reproduce

config file or env var : 
scrapeUri=none

requests for multiple nodes from your prometheus on the same solace prometheus exporter : 
/cloud?scrapeUri=broker1.io:443
/cloud2?scrapeUri=broker1.io:443
/cloud?scrapeUri=broker2.io:443
/cloud2?scrapeUri=broker2.io:443
/brokerha?scrapeUri=solace1.onprem.lan:443
/brokerha2?scrapeUri=solace1.onprem.lan:443
/cluster?scrapeUri=solaceb1.onprem.lan:443
/cluster?scrapeUri=solaceb2.onprem.lan:443
/cluster?scrapeUri=solaceb3.onprem.lan:443

### Solace Broker version

solace cloud, solace gcp, solace on prem

### Solace API

latest

### Solace API version

latest

**Comment by osd42** [comment]
Hello, 

I implemented today on our test environment the new prefetch feature with a node dedicated exporter for the timeout brokerha we have issues with.

I found another argument today: 
- I am using solace prometheus exporters in HA (with dynatrace activegates hosting SPE as docker instance)
- when I configure a solace node dedicated exporter with prefecth, it doubles the overhead due to the config file prefecth always prefetching

If we only prefetch needed scopes based on external requests, it could prevent prefetching unused endpoints. And so it will prefetch data only on the active solace prometheus exporter. So I suppose we could have a prefetch stop rule based on the prefetch interval. If we do not serve a cached data for 3 times the frequency, we could stop prefetch it and wait for a useful request.

Best regards

**Comment by GreenRover** [comment]
As you already start to understand, having prefetching paired with dynamic (per url provided) endpoints is has many pitfalls falls.
This is why it is not provided at the moment. The only possible solution in my options is to provide a preconfigured list of endpoints in configuration.

Another point is:
- if your broker is so large you run in timeouts.
This may harm your broker.
- May be reduce the scrape interval to increase timeouts at prometheus.
- Choose the required endpoints and metrics with care.

What would helps this product is an improvement of documentation.
Please think about the issues that took you time to understand and provide documentation improvements for this. 

But keep always in mind that there are multiple use cases for the exporter:
- my use case: scraping > 300 mostly cloud broker with small foot print (less client and queues but high throughput)
- use case where prefetch was designed for: a small amount of very heavy loaded broker.

**Comment by GreenRover** [comment]
You use case sound similar to mine.
May be have a look at: https://github.com/solacecommunity/solace-prometheus-exporter/tree/master/examples/nginx_reverse_proxy
Here you can play around with: https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout
Another plus is (a very important for me), the monitoring dont need to know about the broker credentials.
And you protect your broker against the redundant querys of ha Prometheus setups.

**Comment by osd42** [comment]
Hello @GreenRover 
Thank you for your answers.

yes the original bug report seems more like a feature request now it is almost in production ! :D

We finally choose the prefetch feature over HA. I did not convince my colleagues that some timeout errors are always better than no HA at all, but they do not want to receive alerts from (Dynatrace) collection issues. So as long as this instance is up, the timeout rate is next to 0% and requests served instantly ! Again, thank you and congratulations for the feature ! 

Yes a list of hosts to replace the string "scrapeURI" could make the trick and reduce the number of solace exporters: one per type of node with a specific config.ini (in my case: brokerha, cluster nodes, solace cloud). But it seems I am using it like you : you share a maximum of exporters, and you create specific exporters only if you have a timeout issue. If this is the good use case I am going to close this issue.

About your nginx reverse proxy, the solution was rejected by Security. In most observability and/or security tools all the GET requests are logged. So we should see nginx rewriting GET requests for SPE, but still MITM GET requests containing clear credentials, nginx logs containing credentials, etc. Prefetch with a reverse proxy seems impossible ATM (same use case as this feature request). We prefered to instanciate 1 SPE per solace cloud (as we cannot change the monitoring username in solace cloud), 1 prefetch SPE per heavy loaded brokerha, 1 shared SPE for all clustered nodes (using GET "&scrapeURI="). This way, we never send any "bad" clear GET request to the SPE prometheus API and only HTTPS POST between SPE and the solace API.

Best regards
